### ISSUE TITLE
RStudio silently converts unicode character in Scripts and Console

### ISSUE THREAD

**Issue by DanteDT** [rstudio/rstudio#10294]
### System details

    RStudio Edition : Desktop
    RStudio Version : RStudio 2021.09.1+372 "Ghost Orchid" Release for Windows
    OS Version      : Win11 21H2 b. 22000.376
    R Version       :  platform       x86_64-w64-mingw32
                            system         x86_64, mingw32             
                            version.string R version 4.1.2 (2021-11-01) Bird Hippie        

### Steps to reproduce the problem
Test using 3 similar but distinct unicode characters:
1. hyphen-minus (-), hex \u002d
1. hyphen (‐), hex \u2010
1. non-breaking hyphen (‑), hex \u2011
```
# Test equivalence of two different hyphen chars
identical("‐", "‑")
identical("\u2010", "\u2011")

# Substitute those 3 for the corresponding number
gsub("\u2010\u2011\u002d", "123", "‐‑-")
```
Paste that code into a new R Script, select all, and "Run". 
Paste same code directly into Console.

### Describe the problem in detail

RStudio produces incorrect results, since it has converted characters somewhere/somehow.
RStudio converts unicode characters, so that 
* identical() mistakenly reports TRUE, and 
* gsub() fails to substitute

This makes related efforts very difficult in RStudio.

### Describe the behavior you expected

Execute same code directly in R's own basic console.
R console for version 4.1.2 (2021-11-01) produces correct results:
```
> # Test equivalence of two different hyphen chars
> identical("‐", "‑")
[1] FALSE
> identical("\u2010", "\u2011")
[1] FALSE
> 
> # Substitute those 3 for the corresponding number
> gsub("\u2010\u2011\u002d", "123", "‐‑-")
[1] "123"
```

- [ x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@DanteDT Thank you for raising the issue! I can't reproduce it, though, with RStudio Desktop 2022.02.0+372 on 
- Windows 10
- MacOS 12.1

Perhaps the issue is do to Windows 11. @DanteDT Do you know by chance if this worked in Windows 10, before upgrading to Windows 11?

**Comment by DanteDT** [comment]
Unfortunately I don't have a Win10 env to test. One additional line confirms that the test chars (various hyphens) are as intended:

```
charToRaw("‐‑-")
gsub("\u2010\u2011\u002d", "123", "‐‑-")
```

My R console results:
```
> charToRaw("‐‑-")
[1] e2 80 90 e2 80 91 2d
> gsub("\u2010\u2011\u002d", "123", "‐‑-")
[1] "123"
```

My RStudio results -- distinct hyphens converted to \u002d
```
> charToRaw("‐‑-")
[1] 2d 2d 2d
> gsub("\u2010\u2011\u002d", "123", "‐‑-")
[1] "---"
```

**Comment by DanteDT** [comment]
I can test older RStudio / R versions in my corporate environment - Win10/Linux.
Ron, your suspicions about Win11 seem like the right direction!

This is what I tested ... I'll follow up a bit later with a Win10 test

```
charToRaw("‐‑‒–—―-")
gsub("\u2010\u2011\u2012\u2013\u2014\u2015\u002d", "1234567", "‐‑‒–—―-")
```

*RStudio Pro:*
```
Version 1.2.1335-1
© 2009-2019 RStudio, Inc.
Build 6 (7b3e38af)
Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:95.0) Gecko/20100101 Firefox/95.0
```

*R versions:*
```
            _                    
platform       x86_64-pc-linux-gnu        
svn rev        80002
version.string R version 4.0.4 (2021-02-15) Lost Library Book
```
and
```
              _                          
platform       x86_64-pc-linux-gnu        
svn rev        76217
version.string R version 3.5.3 (2019-03-11) Great Truth
```

I get correct results for both R versions:

```
> charToRaw("‐‑‒–—―-")
[1] e2 80 90 e2 80 91 e2 80 92 e2 80 93 e2 80 94 e2 80 95 2d
> gsub("\u2010\u2011\u2012\u2013\u2014\u2015\u002d", "1234567", "‐‑‒–—―-")
[1] "1234567"
```

**Comment by DanteDT** [comment]
Update for Win10 - Problems persist, so I no longer suspect Win11.

```
**Win10:** windows 20h2 b. 19042.1415
**RStudio Desktop:** 2021.09.1+372 "Ghost Orchid" Release (8b9ced188245155642d024aa3630363df611088a, 2021-11-08) for Windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36
**R:**  platform x86_64-w64-mingw32, svn rev 81115, Bird Hippie
```

I've used the attached UTF-8 text files (.txt extension, just to be able to attach them here). The original UTF-8 version is **"test-hex.R"**. I opened and "save-as" from several text editors. Except for the native **R Editor**, which corrupts UTF-8 chars with "save as", WinMerge confirms that **all others are identical UTF-8 BOM files**.

* Notepad++: "test-hex-npp.R"
* EditPad:  "test-hex-ep.R"
* VS Code:  "test-hex-vsc.R"
* R Editor:  "test-hex-rEditor.R" - "Save as" corrupts UTF-8 chars, writes a Windows-1252 encoded file
* RStudio:  "test-hex-rStudio.R"

I used Notepad++ (and of course Win10 clipboard) to copy/paste the test code between apps and consoles.

* I detect **no** corruption from this Notepad++/Win10 combination.

I tested 3 actions in both R and RStudio:

1. open and run the test-hex.R code
1. paste into editor, select and run the test code
1. paste directly into console to run the test code

**Expected results for each action:**
```
> charToRaw("‐‑‒–—―-")
 [1] e2 80 90 e2 80 91 e2 80 92 e2 80 93 e2 80 94 e2 80 95 2d
> gsub("\u2010\u2011\u2012\u2013\u2014\u2015\u002d", "1234567", "‐‑‒–—―-")
[1] "1234567"
```

**R results:**
1. as expected
1. as expected
1. as expected

**RStudio results:**
1. failed
1. failed
1. failed

All attachments:

* [test-hex.R.txt](https://github.com/rstudio/rstudio/files/7803993/test-hex.R.txt)
* [test-hex-ep.R.txt](https://github.com/rstudio/rstudio/files/7803994/test-hex-ep.R.txt)
* [test-hex-npp.R.txt](https://github.com/rstudio/rstudio/files/7803995/test-hex-npp.R.txt)
* [test-hex-rEditor.R.txt](https://github.com/rstudio/rstudio/files/7803996/test-hex-rEditor.R.txt)
* [test-hex-rStudio.R.txt](https://github.com/rstudio/rstudio/files/7803997/test-hex-rStudio.R.txt)
* [test-hex-vsc.R.txt](https://github.com/rstudio/rstudio/files/7803998/test-hex-vsc.R.txt)

Screenshots

* ![R-Console-Win10-test-hex](https://user-images.githubusercontent.com/7212088/147979819-09972d30-e18e-44bb-ae12-8617e6b7caa5.png)
* ![RStudio-Win10-test-hex](https://user-images.githubusercontent.com/7212088/147979822-0351840f-d04a-47bc-848a-b497ca1f20eb.png)

**Comment by kevinushey** [comment]
@DanteDT, do you also see this with the development version of R? https://cran.r-project.org/bin/windows/base/rdevel.html

I suspect that R's switching to using the UCRT might allow RStudio to avoid some of the re-encoding of console input done by RStudio (which can occasionally cause transliteration of characters).

I suspect you'll see the same result from RGui (outside of RStudio) if you use something like:

```
source("path/to/script.R", encoding = "UTF-8", echo = TRUE)
```

**Comment by DanteDT** [comment]
Kevin - source() does behave differently and better in R-devel! Details, below.
Ron mentioned a development (?) version of RStudio Desktop, 2022.02.0+372. 
Is there a Win10 installer I could also try?

**4.1.2 (2021-11-01) Bird Hippie**, latest release I started with at top:
```
> source("test-hex.R", encoding="UTF-8", echo=TRUE)
Warning message:
In readLines(file, warn = FALSE) :
  invalid input found on input connection 'test-hex.R'
```

**R-devel** as of today, R Under development (unstable) (2022-01-01 r81419 ucrt)
```
> source("test-hex.R", encoding="UTF-8", echo=TRUE)
> # FILE OPEN
> charToRaw("‐‑‒–—―-")
 [1] e2 80 90 e2 80 91 e2 80 92 e2 80 93 e2 80 94 e2 80 95 2d
> gsub("\u2010\u2011\u2012\u2013\u2014\u2015\u002d", "1234567", "‐‑‒–—―-")
[1] "1234567"
```

**Comment by DanteDT** [comment]
Just to confirm that with the latest release, which addresses some unicode issues, these problems remain.

RStudio 2022.02.0+443 "Prairie Trillium" Release (9f7969398b90468440a501cf065295d9050bb776, 2022-02-16) for Windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36

```
> source("test-hex.R", encoding="UTF-8", echo=TRUE)
Warning message:
In readLines(file, warn = FALSE) :
  invalid input found on input connection 'test-hex.R'
```

**Comment by jcarbaut** [comment]
Similar problem with greek characters. In the following, the characters are correctly stored in the R file, but RStudio "randomly" replaces characters.

Windows 10 (10.0.19044.1706)
R 4.1.2
RStudio 2022.2.0.443

Also tested on RStudio Server on Ubuntu: no issue.

```
u <- c(913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924,
       925, 926, 927, 928, 929, 931, 932, 933, 934, 935, 936, 937)

v <- c(945, 946, 947, 948, 949, 950, 951, 952, 953, 954, 955, 956,
       957, 958, 959, 960, 961, 963, 964, 965, 966, 967, 968, 969)

a <- intToUtf8(u)
b <- intToUtf8(v)

nchar(a) # 24 as expected
nchar(b) # 24 as expected

a <- "ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩ"
b <- "αβγδεζηθικλμνξοπρστυφχψω"
nchar(a) # 23
nchar(b) # 22

Encoding(a)
Encoding(b)

utf8ToInt(a)
# 913   914    71   916   917   918   919    84   921   922   923   924   925
# 926   927   928   929    83   932 14943   935   936    79
utf8ToInt(b)
# 97   223 15165   101   950   951   952   953   954   955   181   957   958
# 959   112   961   115   116 15455   967   968   969
```

**Comment by kevinushey** [comment]
@jcarbaut, do you observe the issue with R 4.2.0?

**Comment by jcarbaut** [comment]
@kevinushey Solved in 4.2.0! I checked on my home computer as I don't have admin privileges on the other one (also Windows 10 and RStudio 2022.02.3 Build 492). Thanks.

**Comment by kevinushey** [comment]
Great to hear! Given that this has been fixed in R as of 4.2.0, I think we can close this issue -- I don't think there's any extra action required on the RStudio side; we'll encourage Windows users to update to R 4.2.0 as soon as possible.
